frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,390 @@
+					let ethereum_address =
+						ethereum_recover(T::Prefix::get(), &reward_account_encoded, &eth_proof)
+							.ok_or(Error::<T>::InvalidProof)?;
+					Result::<_, DispatchError>::Ok(RemoteAccount::Ethereum(ethereum_address))
+				},
+				Proof::RelayChain(relay_account, relay_proof) => {
+					ensure!(
+						verify_relay(
+							T::Prefix::get(),
+							reward_account.clone(),
+							relay_account.clone(),
+							&relay_proof
+						),
+						Error::<T>::InvalidProof
+					);
+					Ok(RemoteAccount::RelayChain(relay_account))
+				},
+			}?;
+			let claimed = Self::do_claim(remote_account.clone(), &reward_account)?;
+			Associations::<T>::insert(reward_account.clone(), remote_account.clone());
+			Self::deposit_event(Event::Associated {
+				remote_account: remote_account.clone(),
+				reward_account: reward_account.clone(),
+			});
+			Self::deposit_event(Event::Claimed { remote_account, reward_account, amount: claimed });
+			Ok(Pays::No.into())
+		}
+
+		/// Claim a reward from the associated reward account.
+		/// A previous call to `associate` should have been made.
+		/// If logic gate pass, no fees are applied.
+		#[pallet::weight(10_000)]
+		pub fn claim(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let reward_account = ensure_signed(origin)?;
+			let remote_account = Associations::<T>::try_get(&reward_account)
+				.map_err(|_| Error::<T>::NotAssociated)?;
+			let claimed = Self::do_claim(remote_account.clone(), &reward_account)?;
+			Self::deposit_event(Event::Claimed { remote_account, reward_account, amount: claimed });
+			Ok(Pays::No.into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Do claim the reward for a given remote account, rewarding the `reward_account`.
+		/// Returns `InvalidProof` if the user is not a contributor or `NothingToClaim` if not
+		/// reward can be claimed yet.
+		fn do_claim(
+			remote_account: RemoteAccountOf<T>,
+			reward_account: &T::AccountId,
+		) -> Result<T::Balance, DispatchError> {
+			ensure!(VestingBlockStart::<T>::exists(), Error::<T>::NotInitialized);
+			Rewards::<T>::try_mutate(remote_account, |reward| {
+				reward
+					.as_mut()
+					.map(|reward| {
+						let upfront_payment = T::InitialPayment::get().mul_floor(reward.total);
+						let should_have_claimed = {
+							let current_block = frame_system::Pallet::<T>::block_number();
+							// Current point in time
+							let vesting_point =
+								current_block.saturating_sub(VestingBlockStart::<T>::get());
+							if vesting_point >= reward.vesting_period {
+								// If the user is claiming when the period is over, he should
+								// probably have already claimed everything.
+								reward.total
+							} else {
+								let vesting_period = VestingBlockStart::<T>::get()
+									.safe_add(&reward.vesting_period)?;
+								let partition = T::VestingPartition::get();
+								// Current window, rounded to previous window.
+								let vesting_window = vesting_point - (vesting_point % partition);
```

Comment:
vesting_period = vesting_end_block = vesting_start_block + period
vesting_point = (current_block - vesting_start_block)
vesting_window = (current_block - vesting_start_block) - (current_block - vesting_start_block) % partition
if current_block is very big, than `vesting_window = current_block`

available_to_claim = vested_reward * current_block / vesting_end_block = available_to_claim grows with each block. nice

Reply:
not sure if i miss something

Reply:
missed, upper rule to return total. huh.